### ISSUE TITLE
Adding neo4j-java-driver-spring-boot-starter prevents Spring Boot app from graceful shutdown

### ISSUE THREAD

**Issue by lilianaziolek-wayfair** [neo4j/neo4j-java-driver-spring-boot-starter#39]
I'm using Neo4j community 4.2.5 started from docker locally as server to connect to.
I've added neo4j-java-driver-spring-boot-starter to an existing app and enabling it (with no other changes, just adding a connection uri and simple query) prevents the app from stopping gracefully. The app is NOT a webapp - just a console application that executes a job and closes.
To reproduce - ideally generate Kotlin Spring Boot project from the spring initializr. I expect the same could happen in Java but I'm using Kotlin.

This starts, prints message and shuts down right after
```kotlin
@SpringBootApplication
class MyApplication : ApplicationRunner {
    override fun run(args: ApplicationArguments?) {
        println("Hello world")
    }
}

fun main(args: Array<String>) {
    runApplication<MyApplication>(*args)
}
```
When I add connection URI in application.yaml and a simple query, without line 1 the app hangs at the end:
```kotlin
@SpringBootApplication
class MyApplication(val driver: Driver) : ApplicationRunner {
    override fun run(args: ApplicationArguments?) {
        println("Hello world")
        val s = driver.session()
        s.use { session -> session.run("MATCH (n) RETURN n LIMIT 3").stream().forEach { println(it) } } //kotlin's equivalent of try-with-resources, autoclosing at the end of the block
        driver.close() //1
    }
}

fun main(args: Array<String>) {
    runApplication<MyApplication>(*args)
}
```
Adding line 1 makes the app close but returns exit code 1:
```
2021-05-13 18:45:34.363  INFO 8648 --- [  restartedMain] c.w.f.s.i.testpckg.MyApplicationKt       : Started MyApplicationKt in 3.156 seconds (JVM running for 5.841)
Hello world
Record<{n: node<0>}>
Record<{n: node<1>}>
Record<{n: node<2>}>
2021-05-13 18:45:35.880  INFO 8648 --- [  restartedMain] org.neo4j.driver.Driver                  : Closing driver instance 2084880078
2021-05-13 18:45:35.883  INFO 8648 --- [  restartedMain] org.neo4j.driver.ConnectionPool          : Closing connection pool towards localhost:7687
2021-05-13 18:45:36.895  INFO 8648 --- [extShutdownHook] o.s.i.endpoint.EventDrivenConsumer       : Removing {logging-channel-adapter:_org.springframework.integration.errorLogger} as a subscriber to the 'errorChannel' channel
2021-05-13 18:45:36.896  INFO 8648 --- [extShutdownHook] o.s.i.channel.PublishSubscribeChannel    : Channel 'application.errorChannel' has 0 subscriber(s).
2021-05-13 18:45:36.896  INFO 8648 --- [extShutdownHook] o.s.i.endpoint.EventDrivenConsumer       : stopped bean '_org.springframework.integration.errorLogger'
2021-05-13 18:45:36.898  INFO 8648 --- [extShutdownHook] o.s.s.c.ThreadPoolTaskScheduler          : Shutting down ExecutorService 'taskScheduler'

Process finished with exit code 1
```

Am I missing something or is this a bug?

**Comment by lilianaziolek-wayfair** [comment]
To further clarify, I am on Spring Boot 2.4 and tried with both `implementation("org.neo4j.driver:neo4j-java-driver:4.2.0")` and `implementation("org.neo4j.driver:neo4j-java-driver-spring-boot-starter:4.2.4.0")` as dependencies and the result is the same - not sure if it's the right place for the issue, if not please let me know where it should be reported.

**Comment by michael-simons** [comment]
Hi @lilianaziolek-wayfair There's little I can do about that :( It's because the Java driver - not this shim here - is creating a background non-daemon thread. If you close it manually, the app will close.

Of course the driver is registered as a auto closable resource with the application and it will be safely stopped when the application is stopped, but that behavior is expected to happen. Please feel free to raise that over there: https://github.com/neo4j/neo4j-java-driver/issues

**Comment by lilianaziolek-wayfair** [comment]
@michael-simons Do you have any more details on where this thread is created / how to get hold of it to stop it?

**Comment by michael-simons** [comment]
It's an internal Netty thread I think… maybe @injectives  or @gjmwoods  can shed a light.

**Comment by michael-simons** [comment]
Coming back to this @lilianaziolek-wayfair: 

I am working on a patch for the driver that demonizes our event loop threads correctly and wether it affects our tests or not. We think that this is pretty much a no-brainer, especially as it will be the default in Netty5 at some day: https://github.com/netty/netty/issues/2832

Thanks for bringing this up!
It's not like that I haven't been annoyed by this behavior beforehand.

**Comment by lilianaziolek-wayfair** [comment]
Wonderful to hear! Thanks and let me know if you need a guinea pig to test anything :)

**Comment by michael-simons** [comment]
This has been addressed in all driver versions (4.1, 4.2 and 4.3). Thank you.

See: https://github.com/neo4j/neo4j-java-driver-spring-boot-starter/releases/tag/4.2.7.0
